File: packages/core-dart/lib/src/routing/extract.dart
import '../address/codes.dart' as codes;
import '../address/codes.dart' show WarningContext;
import '../address/parse.dart';
import '../muxed/decode.dart';
import 'routing_result.dart';
import 'memo.dart';
import 'safe_routing_id.dart';
import 'severity.dart';

/// Extracts deposit routing information from a Stellar payment input.
/// Following the standard priority policy, M-address identifiers take
/// precedence over any provided memo.
///
/// Zero-throw policy: this function never throws for any valid string input.
/// C-addresses, empty destinations, and unrecognized prefixes are returned as
/// a structured [RoutingResult] with either an `INVALID_DESTINATION` warning
/// or a [DestinationError], rather than raising an [ExtractRoutingException].
/// The only remaining exception case is when [RoutingInput.destination] is
/// truly invalid at the type level (which cannot happen in typed Dart code).
///
/// Web safety: routing IDs are resolved through [SafeRoutingId], which
/// parses the canonical decimal **string** exactly and never converts
/// through `int`/JS `Number`. Combined with the `BigInt`-backed
/// [RoutingResult.id], [RoutingResult.idString], and
/// [RoutingResult.safeId] accessors, MEMO_IDs and muxed IDs up to the
/// uint64 ceiling survive Flutter Web without truncation.
///
/// Deprecated synchronous alias of [extractRouting]. For async network checks
/// (Federation, SEP-0029), use [extractRoutingAsync].
///
/// Warnings below [RoutingInput.minSeverityLevel] are filtered out using the
/// shared severity ordering (info = 0, warn = 1, error = 2).
RoutingResult extractRoutingSync(RoutingInput input) {
  return _filterResultBySeverity(_extractRoutingUnfiltered(input), input.minSeverityLevel);
}

/// Re-applies [RoutingInput.minSeverityLevel] to an already-built result.
///
/// Returns [result] unchanged when the threshold is `info` (or unrecognized),
/// which is the common path and avoids re-allocating the warning list.
RoutingResult _filterResultBySeverity(RoutingResult result, String? minSeverityLevel) {
  if (severityWeight(minSeverityLevel) == 0) return result;
  return RoutingResult(
    source: result.source,
    id: result.id,
    destinationBaseAccount: result.destinationBaseAccount,
    destinationError: result.destinationError,
    warnings: filterBySeverity(result.warnings, minSeverityLevel),
  );
}

RoutingResult _extractRoutingUnfiltered(RoutingInput input) {
  final trimmed = input.destination.trim();

  // Empty destination → return structured destinationError (zero-throw policy).
  if (trimmed.isEmpty) {
    return RoutingResult(
      source: RoutingSource.none,
      warnings: [],
      destinationError: DestinationError(
        code: codes.ErrorCode.unknownPrefix,
        message: 'Invalid input: destination must be a non-empty string.',
      ),
    );
  }

  final parsed = parse(input.destination);

  if (parsed.kind == codes.AddressKind.c) {
    return RoutingResult(
      source: RoutingSource.none,
      warnings: [
        for (final w in parsed.warnings)
          RoutingWarning(code: w.code, severity: w.severity, message: w.message),
        RoutingWarning.invalidDestination,
      ],
    );
  }

  if (input.sourceAccount != null && input.sourceAccount!.isNotEmpty) {
    try {
      final source = parse(input.sourceAccount!);
      if (source.kind == codes.AddressKind.c) {
        return RoutingResult(
          source: RoutingSource.none,
          warnings: [RoutingWarning.contractSender],
        );
      }
    } catch (_) {
      // Ignore source account parsing errors for routing extraction
    }
  }

  if (parsed.kind == null) {
    return RoutingResult(
      source: RoutingSource.none,
      warnings: [],
      destinationError: parsed.error != null
          ? DestinationError(
              code: parsed.error!.code,
              message: parsed.error!.message,
            )
          : null,
    );
  }

  final warnings = <RoutingWarning>[];
  for (final w in parsed.warnings) {
    warnings.add(RoutingWarning(
      code: w.code,
      severity: w.severity,
      message: w.message,
    ));
  }

  if (parsed.kind == codes.AddressKind.m) {
    final decoded = MuxedDecoder.decodeMuxedString(parsed.address);
    final baseG = decoded.baseG;
    final muxedId = decoded.id;

    if (input.memoType == 'none') {
      return RoutingResult(
        destinationBaseAccount: baseG,
        id: muxedId,
        source: RoutingSource.muxed,
        warnings: warnings,
      );
    }

    BigInt? routingId;
    RoutingSource routingSource = RoutingSource.none;

    warnings.add(RoutingWarning.memoIgnored);

    if (input.memoType == 'id') {
      final norm = normalizeMemoId(input.memoValue ?? '');
      if (norm.normalized != null) {
        routingId = SafeRoutingId.tryParse(norm.normalized!)?.toBigInt;
        routingSource = RoutingSource.memo;
      } else {
        warnings.add(
          const RoutingWarning(
            code: codes.WarningCode.memoIdInvalidFormat,
            severity: codes.WarningSeverity.warn,
            message: 'MEMO_ID was empty, non-numeric, or exceeded uint64 max.',
          ),
        );
      }
      for (final w in norm.warnings) {
        warnings.add(RoutingWarning(
          code: w.code,
          severity: w.severity,
          message: w.message,
        ));
      }
    } else if (input.memoType == 'text' && input.memoValue != null) {
      final norm = normalizeMemoTextId(input.memoValue!);
      if (norm.normalized != null) {
        routingId = SafeRoutingId.tryParse(norm.normalized!)?.toBigInt;
        routingSource = RoutingSource.memo;
      } else {
        warnings.add(
          const RoutingWarning(
            code: codes.WarningCode.memoTextUnroutable,
            severity: codes.WarningSeverity.warn,
            message: 'MEMO_TEXT was not a valid numeric uint64.',
          ),
        );
      }
      for (final w in norm.warnings) {
        warnings.add(RoutingWarning(
          code: w.code,
          severity: w.severity,
          message: w.message,
        ));
      }
    } else if (input.memoType == 'hash' || input.memoType == 'return') {
      warnings.add(
        RoutingWarning(
          code: codes.WarningCode.unsupportedMemoType,
          severity: codes.WarningSeverity.warn,
          message: 'Memo type ${input.memoType} is not supported for routing.',
          context: WarningContext(memoType: input.memoType),
        ),
      );
    } else {
      warnings.add(
        const RoutingWarning(
          code: codes.WarningCode.unsupportedMemoType,
          severity: codes.WarningSeverity.warn,
          message: 'Unrecognized memo type: unknown',
          context: WarningContext(memoType: 'unknown'),
        ),
      );
    }

    return RoutingResult(
      destinationBaseAccount: baseG,
      id: routingId,
      source: routingSource,
      warnings: warnings,
    );
  }

  BigInt? routingId;
  RoutingSource routingSource = RoutingSource.none;

  if (input.memoType == 'id') {
    final norm = normalizeMemoId(input.memoValue ?? '');
    if (norm.normalized != null) {
      routingId = SafeRoutingId.tryParse(norm.normalized!)?.toBigInt;
      routingSource = RoutingSource.memo;
    } else {
      warnings.add(
        const RoutingWarning(
          code: codes.WarningCode.memoIdInvalidFormat,
          severity: codes.WarningSeverity.warn,
          message: 'MEMO_ID was empty, non-numeric, or exceeded uint64 max.',
        ),
      );
    }
    for (final w in norm.warnings) {
      warnings.add(RoutingWarning(
        code: w.code,
        severity: w.severity,
        message: w.message,
      ));
    }
  } else if (input.memoType == 'text' && input.memoValue != null) {
    final norm = normalizeMemoTextId(input.memoValue!);
    if (norm.normalized != null) {
      routingId = SafeRoutingId.tryParse(norm.normalized!)?.toBigInt;
      routingSource = RoutingSource.memo;
    } else {
      warnings.add(
        const RoutingWarning(
          code: codes.WarningCode.memoTextUnroutable,
          severity: codes.WarningSeverity.warn,
          message: 'MEMO_TEXT was not a valid numeric uint64.',
        ),
      );
    }
    for (final w in norm.warnings) {
      warnings.add(RoutingWarning(
        code: w.code,
        severity: w.severity,
        message: w.message,
      ));
    }
  } else if (input.memoType == 'hash' || input.memoType == 'return') {
    warnings.add(
      RoutingWarning(
        code: codes.WarningCode.unsupportedMemoType,
        severity: codes.WarningSeverity.warn,
        message: 'Memo type ${input.memoType} is not supported for routing.',
        context: WarningContext(memoType: input.memoType),
      ),
    );
  } else if (input.memoType != 'none') {
    warnings.add(
      const RoutingWarning(
        code: codes.WarningCode.unsupportedMemoType,
        severity: codes.WarningSeverity.warn,
        message: 'Unrecognized memo type: unknown',
        context: WarningContext(memoType: 'unknown'),
      ),
    );
  }

  return RoutingResult(
    destinationBaseAccount: parsed.address,
    id: routingId,
    source: routingSource,
    warnings: warnings,
  );
}

/// Looks up whether [baseAccount] requires a memo on incoming payments
/// (SEP-0029, the `config.memo_required` account data entry).
///
/// Supplied by the caller so this package stays free of network code.
typedef MemoRequirementFetcher = Future<bool> Function(String baseAccount);

/// Extracts deposit routing information, optionally checking the
/// destination's SEP-0029 memo requirement.
///
/// Runs [extractRouting] first. If [fetchMemoRequirement] is given and
/// the result has a destination account but no routing ID, the fetcher is
/// called; when it returns `true`, [RoutingWarning.missingRequiredMemo] is
/// appended. Fetch failures fail open: the synchronous result is returned
/// unchanged.
///
/// ```dart
/// final result = await extractRoutingAsync(
///   RoutingInput(
///     destination: 'GAYCUYT553C5LHVE2XPW5GMEJT4BXGM7AHMJWLAPZP53KJO7EIQADRSI',
///     memoType: 'none',
///   ),
///   fetchMemoRequirement: (account) async => horizon.requiresMemo(account),
/// );
/// if (result.warnings.contains(RoutingWarning.missingRequiredMemo)) {
///   // Hold the deposit for manual review.
/// }
/// ```
Future<RoutingResult> extractRoutingAsync(
  RoutingInput input, {
  MemoRequirementFetcher? fetchMemoRequirement,
}) async {
  final result = extractRouting(input);
  if (fetchMemoRequirement == null ||
      result.destinationBaseAccount == null ||
      result.id != null ||
      result.destinationError != null) {
    return result;
  }

  try {
    if (await fetchMemoRequirement(result.destinationBaseAccount!)) {
      final withMemoWarning = RoutingResult(
        source: result.source,
        id: result.id,
        destinationBaseAccount: result.destinationBaseAccount,
        destinationError: result.destinationError,
        warnings: [...result.warnings, RoutingWarning.missingRequiredMemo],
      );
      return _filterResultBySeverity(withMemoWarning, input.minSeverityLevel);
    }
  } catch (_) {
    // Network/configuration failures must not change the synchronous result.
  }
  return result;
}

File: packages/core-go/routing/extract.go
package routing

import (
	"context"
	"strconv"
	"strings"

	"github.com/REDISHFISH/BLUEWHALE/packages/core-go/address"
)

// normalizeUnsupportedMemoType canonicalizes a memo type string by lower-casing it
// and stripping underscores and hyphens, then maps it to a known unsupported type.
// Uses allocation-free byte comparisons: exact-match fast paths for canonical types, then a folded compare.
func normalizeUnsupportedMemoType(memoType string) string {
	switch memoType {
	case "hash", "return":
		return memoType
	}

	switch memoType {
	case "none", "id", "text":
		return ""
	}

	switch {
	case foldedMemoTypeEquals(memoType, "memohash"):
		return "hash"
	case foldedMemoTypeEquals(memoType, "memoreturn"):
		return "return"
	default:
		return ""
	}
}

// ExtractRouting identifies the deposit routing destination and identifier from a Stellar
// payment input. It implements the standard priority policy where M-address identifiers
// take precedence over any provided memo. Returns a RoutingResult with the decoded
// state and applicable warnings, filtered by input.MinSeverityLevel.
func ExtractRouting(input RoutingInput) RoutingResult {
	result := extractRouting(input)
	result.Warnings = FilterBySeverity(result.Warnings, input.MinSeverityLevel)
	return result
}

func extractRouting(input RoutingInput) RoutingResult {
	if input.SourceAccount != "" {
		source, err := address.Parse(input.SourceAccount)
		if err == nil && source.Kind == address.KindC {
			return RoutingResult{
				RoutingSource: "none",
				Warnings: []address.Warning{{
					Code:     address.WarnContractSenderDetected,
					Severity: "info",
					Message:  "Contract source detected. Routing state cleared.",
				}},
			}
		}
	}

	parsed, err := address.Parse(input.Destination)
	if err != nil {
		return RoutingResult{
			RoutingSource: "none",
			Warnings:      []address.Warning{},
			DestinationError: &DestinationError{
				Code:    address.ErrUnknownPrefix,
				Message: err.Error(),
			},
		}
	}

	if parsed.Kind == address.KindC {
		return RoutingResult{
			RoutingSource: "none",
			Warnings: []address.Warning{{
				Code:     address.WarnInvalidDestination,
				Severity: "error",
				Message:  "C address is not a valid destination",
				Context: &address.WarningContext{
					DestinationKind: "C",
				},
			}},
		}
	}

	if parsed.Kind == address.KindM {
		// address.Parse already decoded the muxed payload; reuse it rather
		// than decoding the M-address a second time.
		baseG, id := parsed.BaseG, parsed.MuxedID

		// Pre-allocate with capacity for existing warnings plus at most one more.
		warnings := make([]address.Warning, 0, len(parsed.Warnings)+1)
		warnings = append(warnings, parsed.Warnings...)
		memoValue := stringValue(input.MemoValue)

		// isAllDigits replaces the regex match to avoid heap allocation.
		if input.MemoType == "id" || (input.MemoType == "text" && isAllDigits(memoValue)) {
			warnings = append(warnings, address.Warning{
				Code:     address.WarnMemoPresentWithMuxed,
				Severity: "warn",
				Message:  "Routing ID found in both M-address and Memo. M-address ID takes precedence.",
			})
		} else if input.MemoType != "none" {
			warnings = append(warnings, address.Warning{
				Code:     address.WarnMemoIgnoredForMuxed,
				Severity: "info",
				Message:  "Memo present with M-address. Any potential routing ID in memo is ignored.",
			})
		}

		return RoutingResult{
			DestinationBaseAccount: baseG,
			RoutingID:              NewRoutingID(strconv.FormatUint(id, 10)),
			RoutingSource:          "muxed",
			Warnings:               warnings,
		}
	}

	var routingID *RoutingID
	routingSource := "none"
	// Pre-allocate with capacity for existing address warnings plus at most two memo warnings.
	warnings := make([]address.Warning, 0, len(parsed.Warnings)+2)
	warnings = append(warnings, parsed.Warnings...)
	memoValue := stringValue(input.MemoValue)

	if input.MemoType == "id" {
		norm := NormalizeMemoTextID(memoValue)
		if norm.Normalized != "" {
			routingID = NewRoutingID(norm.Normalized)
			routingSource = "memo"
		}
		warnings = append(warnings, norm.Warnings...)

		if norm.Normalized == "" {
			warnings = append(warnings, address.Warning{
				Code:     address.WarnMemoIDInvalidFormat,
				Severity: "warn",
				Message:  "MEMO_ID was empty, non-numeric, or exceeded uint64 max.",
			})
		}
	} else if input.MemoType == "text" && memoValue != "" {
		norm := NormalizeMemoTextID(memoValue)
		if norm.Normalized != "" {
			routingID = NewRoutingID(norm.Normalized)
			routingSource = "memo"
			warnings = append(warnings, norm.Warnings...)
		} else {
			warnings = append(warnings, address.Warning{
				Code:     address.WarnMemoTextUnroutable,
				Severity: "warn",
				Message:  "MEMO_TEXT was not a valid numeric uint64.",
			})
		}
	} else if unsupportedMemoType := normalizeUnsupportedMemoType(input.MemoType); unsupportedMemoType != "" {
		// Use pre-computed string literals for known memo types to avoid string concatenation.
		var msg string
		switch unsupportedMemoType {
		case "hash":
			msg = "Memo type hash is not supported for routing."
		case "return":
			msg = "Memo type return is not supported for routing."
		}
		warnings = append(warnings, address.Warning{
			Code:     address.WarnUnsupportedMemoType,
			Severity: "warn",
			Message:  msg,
			Context: &address.WarningContext{
				MemoType: unsupportedMemoType,
			},
		})
	} else if input.MemoType != "none" {
		// Use strings.Builder to avoid the two intermediate allocations from "prefix" + var concatenation.
		var sb strings.Builder
		sb.Grow(len("Unrecognized memo type: ") + len(input.MemoType))
		sb.WriteString("Unrecognized memo type: ")
		sb.WriteString(input.MemoType)
		warnings = append(warnings, address.Warning{
			Code:     address.WarnUnsupportedMemoType,
			Severity: "warn",
			Message:  sb.String(),
			Context: &address.WarningContext{
				MemoType: "unknown",
			},
		})
	}

	return RoutingResult{
		DestinationBaseAccount: parsed.Raw,
		RoutingID:              routingID,
		RoutingSource:          routingSource,
		Warnings:               warnings,
	}
}

// MemoRequirementFetcher retrieves whether a destination account requires a
// routing memo. Implementations can use Horizon, an indexer, or a cached source.
//
// Deprecated: Use [ContextMemoRequirementFetcher] and
// [ExtractRoutingWithContext] instead, which accept a [context.Context] for
// cancellation and deadline propagation.
type MemoRequirementFetcher func(baseAccount string) (bool, error)

// ContextMemoRequirementFetcher is the context-aware form of
// [MemoRequirementFetcher]. Implementations should honour ctx cancellation so
// an in-flight Horizon or indexer lookup does not outlive the caller's
// deadline.
type ContextMemoRequirementFetcher func(ctx context.Context, baseAccount string) (bool, error)

// ExtractRoutingWithContext performs normal routing extraction and
// optionally adds the SEP-0029 error when a classic destination requires a
// memo but no routing ID was supplied. Fetch failures fail open so callers
// retain the result of the synchronous parser.
//
// ctx is forwarded to fetch, allowing a cancelled or expired context to abort
// the lookup; a fetcher that returns an error contributes no warning.
func ExtractRoutingWithContext(ctx context.Context, input RoutingInput, fetch ContextMemoRequirementFetcher) RoutingResult {
	result := ExtractRouting(input)
	if fetch == nil || result.DestinationBaseAccount == "" || result.RoutingID != nil || result.DestinationError != nil {
		return result
	}
	if err := ctx.Err(); err != nil {
		return result
	}
	required, err := fetch(ctx, result.DestinationBaseAccount)
	if err == nil && required {
		result.Warnings = append(result.Warnings, address.Warning{
			Code:     address.WarnMissingRequiredMemo,
			Severity: "error",
			Message:  "Destination account requires a memo, but no routing ID was provided.",
		})
	}
	return result
}

// ExtractRoutingWithMemoRequirement performs normal routing extraction and
// optionally adds the SEP-0029 error when a classic destination requires a
// memo but no routing ID was supplied. Fetch failures fail open so callers
// retain the result of the synchronous parser.
//
// Deprecated: Use [ExtractRoutingWithContext] with a [ContextMemoRequirementFetcher]
// to enable request cancellation and deadline propagation.
func ExtractRoutingWithMemoRequirement(input RoutingInput, fetch MemoRequirementFetcher) RoutingResult {
	if fetch == nil {
		return ExtractRoutingWithContext(context.Background(), input, nil)
	}
	// Wrap the legacy fetcher so it satisfies ContextMemoRequirementFetcher.
	// The context is intentionally not forwarded to the wrapped function
	// (it has no context parameter), but callers can migrate to
	// ExtractRoutingWithContext when they are ready.
	return ExtractRoutingWithContext(context.Background(), input, func(_ context.Context, baseAccount string) (bool, error) {
		return fetch(baseAccount)
	})
}

func stringValue(s string) string {
	return s
}
